python/ray/autoscaler/aws/node_provider.py

```diff
@@ -40,15 +44,55 @@ def __init__(self, provider_config, cluster_name):
         # Try availability zones round-robin, starting from random offset
         self.subnet_idx = random.randint(0, 100)
 
+        self.tag_cache = {}  # Tags that we believe to actually be on EC2.
+        self.tag_cache_pending = {}  # Tags that we will soon upload.
+        self.tag_cache_lock = threading.Lock()
+        self.tag_cache_update_event = threading.Event()
+        self.tag_cache_kill_event = threading.Event()
+        self.tag_update_thread = threading.Thread(
+            target=self._node_tag_update_loop
+        )
+        self.tag_update_thread.start()  # TODO: monitor this?
+
         # Cache of node objects from the last nodes() call. This avoids
         # excessive DescribeInstances requests.
         self.cached_nodes = {}
 
-        # Cache of ip lookups. We assume IPs never change once assigned.
-        self.internal_ip_cache = {}
-        self.external_ip_cache = {}
+    def _node_tag_update_loop(self):
```

Comment:
Can you add a comment describing the purpose of this loop?

Reply:
Sure.

The idea is that the tags should be updated asynchronously.

As long as the autoscaler itself knows the status, and a user looking at AWS or performing API calls is eventually made consistent (I choose ~5s as the acceptable lag here), we don't need to immediately update tags (because then we hit AWS with a ton of calls, and with rate limiting, boto limits and latency, it kills performance).

I'll put together a docstring for it.